src/modules/flow/lsm9ds0/lsm9ds0.c

```diff
@@ -0,0 +1,498 @@
+    }
+
+    mdata->init = true;
+}
+
+static inline void
+_init_gyro_queues(struct lsm9ds0_gyro_data *mdata, uint8_t scale)
+{
+    struct sol_i2c_op *op;
+
+    sol_vector_init(&mdata->read_q, sizeof(struct sol_i2c_op));
+    op = sol_vector_append_n(&mdata->read_q, 6);
+
+    // Gyro X Y Z
+    op[0].type = SOL_I2C_READ;
+    op[0].reg = OUT_X_L_G;
+    op[1].type = SOL_I2C_READ;
+    op[1].reg = OUT_X_H_G;
+
+    op[2].type = SOL_I2C_READ;
+    op[2].reg = OUT_Y_L_G;
+    op[3].type = SOL_I2C_READ;
+    op[3].reg = OUT_Y_H_G;
+
+    op[4].type = SOL_I2C_READ;
+    op[4].reg = OUT_Z_L_G;
+    op[5].type = SOL_I2C_READ;
+    op[5].reg = OUT_Z_H_G;
+
+    // Init
+    sol_vector_init(&mdata->init_q, sizeof(struct sol_i2c_op));
+    op = sol_vector_append_n(&mdata->init_q, 5);
+
+    op[0].type = SOL_I2C_WRITE;
+    op[0].reg = CTRL_REG1_G;
+    op[0].value = 0x0F;
+
+    op[1].type = SOL_I2C_WRITE;
+    op[1].reg = CTRL_REG2_G;
+    op[1].value = 0x00;
+
+    op[2].type = SOL_I2C_WRITE;
+    op[2].reg = CTRL_REG4_G;
+    op[2].value = scale;
+
+    op[3].type = SOL_I2C_WRITE;
+    op[3].reg = CTRL_REG5_G;
+    op[3].value = 0x10;
+
+    op[4].type = SOL_I2C_WRITE;
+    op[4].reg = FIFO_CTRL_REG;
+    op[4].value = 0x00;
+}
+
+static int
+lsm9ds0_gyro_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    uint8_t scale_reg = 0x00;
+    struct lsm9ds0_gyro_data *mdata = data;
+    const struct sol_flow_node_type_lsm9ds0_gyroscope_options *opts =
+        (const struct sol_flow_node_type_lsm9ds0_gyroscope_options *)options;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(opts,
+        SOL_FLOW_NODE_TYPE_LSM9DS0_GYROSCOPE_OPTIONS_API_VERSION, -EINVAL);
+
+    mdata->node = node;
+    mdata->init = false;
+
+    mdata->i2c = sol_i2c_open(opts->i2c_bus, SOL_I2C_SPEED_10KBIT);
+    SOL_NULL_CHECK_MSG(mdata->i2c, -EINVAL, "Failed to open i2c bus: %u", opts->i2c_bus);
+
+    mdata->addr = opts->i2c_addr;
+    mdata->scale = opts->scale;
+
+    if (mdata->scale == 2000) {
+        scale_reg = 0x0C;
+    } else if (mdata->scale == 500) {
+        scale_reg = 0x04;
+    } else if (mdata->scale != 245) {
+        SOL_WRN("Invalid Scale option. Using 245 dps. Valid options are: 245, 500, 2000.");
+        mdata->scale = 245;
+    }
+
+    _init_gyro_queues(mdata, scale_reg);
+
+    mdata->pending = sol_i2c_dispatcher_add_op_set(mdata->i2c, mdata->addr, &mdata->init_q,
+        _init_gyro_done, mdata, 0);
+
+    if (!mdata->pending) {
+        SOL_ERR("Couldn't initialize LSM9DS0 Gyroscope.");
+        return -1;
+    }
+
+    return 0;
+}
+
+static void
+lsm9ds0_gyro_close(struct sol_flow_node *node, void *data)
+{
+    struct lsm9ds0_gyro_data *mdata = data;
+
+    if (mdata->i2c) {
+        if (mdata->pending)
+            sol_i2c_dispatcher_remove_op_set(mdata->i2c, mdata->pending);
+        sol_i2c_close(mdata->i2c);
+    }
+
+    sol_vector_clear(&mdata->init_q);
+    sol_vector_clear(&mdata->read_q);
```

Comment:
Why not keep one queue alone? You use one for init, and clear it on the done callback, then use the same vector for all the read queues you may need during the lifetime of the node. Same for the other one below.